File: surface/widgets/swiftui/components/widgets/concepts/process-automation/ExpressionToggleInput.swift
import SwiftUI

enum ExpressionToggleInputWidgetState {
    case fixed, expression, autocompleting
}

enum ExpressionToggleInputEvent {
    case toggle, input, showAC, select, dismiss
}

func expressionToggleInputReduce(state: ExpressionToggleInputWidgetState, event: ExpressionToggleInputEvent) -> ExpressionToggleInputWidgetState {
    switch state {
    case .fixed:
        if event == .toggle { return .expression }
        return state
    case .expression:
        if event == .toggle { return .fixed }
        if event == .showAC { return .autocompleting }
        return state
    case .autocompleting:
        if event == .select { return .expression }
        if event == .dismiss { return .expression }
        return state
    }
}

struct ExpressionToggleInputView: View {
    var value: String
    var mode: String
    var fieldType: String = "text" // "text", "number", "boolean", "object"
    var variables: [String] = []
    var expression: String = ""
    var previewValue: String? = nil
    var expressionValid: Bool? = nil
    var onChange: ((String) -> Void)? = nil
    var onExpressionChange: ((String) -> Void)? = nil
    var onToggleMode: ((String) -> Void)? = nil

    @State private var widgetState: ExpressionToggleInputWidgetState = .fixed
    @State private var fixedValue: String = ""
    @State private var expressionValue: String = ""
    @State private var acQuery: String = ""
    @State private var acIndex: Int = 0

    private var isExpressionMode: Bool { widgetState != .fixed }

    private var suggestions: [String] {
        if acQuery.isEmpty { return variables }
        let q = acQuery.lowercased()
        return variables.filter { $0.lowercased().contains(q) }
    }

    private func handleToggle() {
        let newMode: String = widgetState == .fixed ? "expression" : "fixed"
        widgetState = expressionToggleInputReduce(state: widgetState, event: .toggle)
        onToggleMode?(newMode)
    }

    private func handleFixedChange(_ newValue: String) {
        fixedValue = newValue
        onChange?(newValue)
    }

    private func handleExpressionChange(_ newExpr: String) {
        expressionValue = newExpr
        onExpressionChange?(newExpr)

        let parts = newExpr.components(separatedBy: CharacterSet(charactersIn: " ()+\\-*/,"))
        let lastWord = parts.last ?? ""
        if !lastWord.isEmpty && variables.contains(where: { $0.lowercased().hasPrefix(lastWord.lowercased()) }) {
            acQuery = lastWord
            acIndex = 0
            widgetState = expressionToggleInputReduce(state: widgetState, event: .showAC)
        }
    }

    private func selectSuggestion(_ variable: String) {
        let parts = expressionValue.components(separatedBy: CharacterSet(charactersIn: " ()+\\-*/,"))
        let lastPart = parts.last ?? ""
        let newExpr = String(expressionValue.dropLast(lastPart.count)) + variable
        expressionValue = newExpr
        onExpressionChange?(newExpr)
        widgetState = expressionToggleInputReduce(state: widgetState, event: .select)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            // Mode toggle
            Button(isExpressionMode ? "Expression" : "Fixed") {
                handleToggle()
            }
            .accessibilityLabel("Expression mode")
            .accessibilityValue(isExpressionMode ? "on" : "off")
            .accessibilityRole(.toggleButton)

            // Fixed value input
            if !isExpressionMode {
                switch fieldType {
                case "boolean":
                    Toggle(fixedValue == "true" ? "true" : "false", isOn: Binding(
                        get: { fixedValue == "true" },
                        set: { handleFixedChange(String($0)) }
                    ))
                    .accessibilityLabel("Fixed boolean value")

                case "number":
                    TextField("Number", text: Binding(
                        get: { fixedValue },
                        set: { handleFixedChange($0) }
                    ))
                    .textFieldStyle(.roundedBorder)
                    .accessibilityLabel("Fixed number value")

                case "object":
                    TextEditor(text: Binding(
                        get: { fixedValue },
                        set: { handleFixedChange($0) }
                    ))
                    .font(.system(.body, design: .monospaced))
                    .frame(minHeight: 80)
                    .overlay(RoundedRectangle(cornerRadius: 4).stroke(Color.gray.opacity(0.3), lineWidth: 1))
                    .accessibilityLabel("Fixed object value (JSON)")

                default:
                    TextField("Value", text: Binding(
                        get: { fixedValue },
                        set: { handleFixedChange($0) }
                    ))
                    .textFieldStyle(.roundedBorder)
                    .accessibilityLabel("Fixed text value")
                }
            }

            // Expression editor
            if isExpressionMode {
                TextEditor(text: Binding(
                    get: { expressionValue },
                    set: { handleExpressionChange($0) }
                ))
                .font(.system(.body, design: .monospaced))
                .frame(minHeight: 60)
                .overlay(
                    RoundedRectangle(cornerRadius: 4)
                        .stroke(expressionValid == false ? Color.red : Color.gray.opacity(0.3), lineWidth: 1)
                )
                .accessibilityLabel("Expression editor")
            }

            // Autocomplete dropdown
            if widgetState == .autocompleting {
                VStack(alignment: .leading, spacing: 0) {
                    if suggestions.isEmpty {
                        Text("No matching variables")
                            .font(.caption)
                            .foregroundColor(.secondary)
                            .padding(8)
                    } else {
                        ForEach(Array(suggestions.enumerated()), id: \.element) { index, variable in
                            Text(variable)
                                .font(.system(.body, design: .monospaced))
                                .padding(.horizontal, 8)
                                .padding(.vertical, 4)
                                .frame(maxWidth: .infinity, alignment: .leading)
                                .background(acIndex == index ? Color.accentColor.opacity(0.15) : Color.clear)
                                .onTapGesture { selectSuggestion(variable) }
                                .onHover { hovering in
                                    if hovering { acIndex = index }
                                }
                                .accessibilityLabel(variable)
                                .accessibilityValue(acIndex == index ? "focused" : "")
                        }
                    }
                }
                .background(Color(.systemBackground))
                .overlay(RoundedRectangle(cornerRadius: 4).stroke(Color.gray.opacity(0.3), lineWidth: 1))
                .accessibilityLabel("Variable suggestions")
            }

            // Live preview
            if isExpressionMode {
                if let preview = previewValue {
                    Text(preview)
                        .font(.caption)
                        .foregroundColor(expressionValid == false ? .red : .green)
                        .accessibilityLabel("Preview: \(preview)")
                } else if !expressionValue.isEmpty {
                    Text("Enter expression to preview")
                        .font(.caption)
                        .foregroundColor(.secondary)
                }
            }
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Expression toggle input")
        .onAppear {
            fixedValue = value
            expressionValue = expression
        }
    }
}

File: surface/widgets/nativescript/components/widgets/concepts/process-automation/ExpressionToggleInput.ts
import {
  StackLayout,
  Label,
  Button,
  TextField,
  TextView,
  Switch,
} from '@nativescript/core';

export type ExpressionToggleInputState = 'fixed' | 'expression' | 'autocompleting';
export type ExpressionToggleInputEvent =
  | { type: 'TOGGLE' }
  | { type: 'INPUT'; value?: string }
  | { type: 'SHOW_AC' }
  | { type: 'SELECT'; variable?: string }
  | { type: 'DISMISS' };

export function expressionToggleInputReducer(state: ExpressionToggleInputState, event: ExpressionToggleInputEvent): ExpressionToggleInputState {
  switch (state) {
    case 'fixed':
      if (event.type === 'TOGGLE') return 'expression';
      if (event.type === 'INPUT') return 'fixed';
      return state;
    case 'expression':
      if (event.type === 'TOGGLE') return 'fixed';
      if (event.type === 'INPUT') return 'expression';
      if (event.type === 'SHOW_AC') return 'autocompleting';
      return state;
    case 'autocompleting':
      if (event.type === 'SELECT') return 'expression';
      if (event.type === 'DISMISS') return 'expression';
      return state;
    default:
      return state;
  }
}

export interface ExpressionToggleInputProps {
  value: string;
  mode: string;
  fieldType?: 'text' | 'number' | 'boolean' | 'object';
  variables?: string[];
  expression?: string;
  previewValue?: string;
  expressionValid?: boolean;
  onChange?: (value: string) => void;
  onExpressionChange?: (expression: string) => void;
  onToggleMode?: (mode: 'fixed' | 'expression') => void;
}

export function createExpressionToggleInput(props: ExpressionToggleInputProps): {
  view: StackLayout;
  dispose: () => void;
} {
  let state: ExpressionToggleInputState = 'fixed';
  let fixedValue = props.value;
  let expressionValue = props.expression ?? '';
  let acIndex = 0;
  const variables = props.variables ?? [];
  const disposers: (() => void)[] = [];

  function send(event: ExpressionToggleInputEvent) {
    state = expressionToggleInputReducer(state, event);
    update();
  }

  const root = new StackLayout();
  root.className = 'expression-toggle-input';
  root.automationText = 'Expression toggle input';

  // Mode toggle
  const modeToggle = new Button();
  modeToggle.text = 'Fixed';
  const modeToggleCb = () => {
    const newMode = state === 'fixed' ? 'expression' : 'fixed';
    send({ type: 'TOGGLE' });
    props.onToggleMode?.(newMode as 'fixed' | 'expression');
  };
  modeToggle.on('tap', modeToggleCb);
  disposers.push(() => modeToggle.off('tap', modeToggleCb));
  root.addChild(modeToggle);

  // Fixed input area
  const fixedArea = new StackLayout();
  root.addChild(fixedArea);

  // Expression area
  const exprArea = new StackLayout();
  root.addChild(exprArea);

  // Autocomplete list
  const acArea = new StackLayout();
  acArea.className = 'autocomplete';
  root.addChild(acArea);

  // Preview
  const previewLbl = new Label();
  previewLbl.fontSize = 12;
  previewLbl.marginTop = 4;
  root.addChild(previewLbl);

  function getSuggestions(query: string): string[] {
    if (!query) return variables;
    const q = query.toLowerCase();
    return variables.filter((v) => v.toLowerCase().includes(q));
  }

  function update() {
    const isExpr = state !== 'fixed';
    modeToggle.text = isExpr ? 'Expression' : 'Fixed';

    // Fixed area
    fixedArea.removeChildren();
    fixedArea.visibility = !isExpr ? 'visible' : 'collapsed';
    if (!isExpr) {
      const fieldType = props.fieldType ?? 'text';
      if (fieldType === 'boolean') {
        const boolRow = new StackLayout();
        boolRow.orientation = 'horizontal';
        const sw = new Switch();
        sw.checked = fixedValue === 'true';
        sw.on('checkedChange', () => {
          fixedValue = String(sw.checked);
          send({ type: 'INPUT', value: fixedValue });
          props.onChange?.(fixedValue);
        });
        boolRow.addChild(sw);
        const boolLbl = new Label();
        boolLbl.text = fixedValue === 'true' ? 'true' : 'false';
        boolLbl.marginLeft = 8;
        boolRow.addChild(boolLbl);
        fixedArea.addChild(boolRow);
      } else if (fieldType === 'object') {
        const objInput = new TextView();
        objInput.text = fixedValue;
        objInput.hint = 'JSON value';
        objInput.on('textChange', () => {
          fixedValue = objInput.text;
          send({ type: 'INPUT', value: fixedValue });
          props.onChange?.(fixedValue);
        });
        fixedArea.addChild(objInput);
      } else {
        const tf = new TextField();
        tf.text = fixedValue;
        tf.keyboardType = fieldType === 'number' ? 'number' : 'text';
        tf.on('textChange', () => {
          fixedValue = tf.text;
          send({ type: 'INPUT', value: fixedValue });
          props.onChange?.(fixedValue);
        });
        fixedArea.addChild(tf);
      }
    }

    // Expression area
    exprArea.removeChildren();
    exprArea.visibility = isExpr ? 'visible' : 'collapsed';
    if (isExpr) {
      const exprInput = new TextView();
      exprInput.text = expressionValue;
      exprInput.hint = 'Enter expression...';
      exprInput.on('textChange', () => {
        expressionValue = exprInput.text;
        send({ type: 'INPUT', value: expressionValue });
        props.onExpressionChange?.(expressionValue);

        const lastWord = expressionValue.split(/[\s()+\-*/,]+/).pop() ?? '';
        if (lastWord.length > 0 && variables.some((v) => v.toLowerCase().startsWith(lastWord.toLowerCase()))) {
          acIndex = 0;
          send({ type: 'SHOW_AC' });
        }
      });
      exprArea.addChild(exprInput);
    }

    // Autocomplete
    acArea.removeChildren();
    acArea.visibility = state === 'autocompleting' ? 'visible' : 'collapsed';
    if (state === 'autocompleting') {
      const lastWord = expressionValue.split(/[\s()+\-*/,]+/).pop() ?? '';
      const suggestions = getSuggestions(lastWord);
      if (suggestions.length === 0) {
        const noMatch = new Label();
        noMatch.text = 'No matching variables';
        noMatch.fontSize = 12;
        acArea.addChild(noMatch);
      } else {
        suggestions.forEach((variable, idx) => {
          const item = new Button();
          item.text = variable;
          item.className = idx === acIndex ? 'ac-focused' : 'ac-item';
          item.on('tap', () => {
            const parts = expressionValue.split(/[\s()+\-*/,]+/);
            const lastPart = parts[parts.length - 1] ?? '';
            expressionValue = expressionValue.slice(0, expressionValue.length - lastPart.length) + variable;
            props.onExpressionChange?.(expressionValue);
            send({ type: 'SELECT', variable });
          });
          acArea.addChild(item);
        });
      }
    }

    // Preview
    if (isExpr && props.previewValue !== undefined) {
      previewLbl.visibility = 'visible';
      previewLbl.text = props.previewValue;
      previewLbl.className = props.expressionValid !== false ? 'preview-valid' : 'preview-invalid';
    } else if (isExpr && expressionValue) {
      previewLbl.visibility = 'visible';
      previewLbl.text = 'Enter expression to preview';
    } else {
      previewLbl.visibility = 'collapsed';
    }
  }

  update();

  return {
    view: root,
    dispose() {
      disposers.forEach((d) => d());
    },
  };
}

export default createExpressionToggleInput;
